source/common/router/scoped_rds.cc

```diff
@@ -100,71 +108,182 @@ ScopedRdsConfigSubscription::ScopedRdsConfigSubscription(
   });
 }
 
+ScopedRdsConfigSubscription::~ScopedRdsConfigSubscription() { cleanup_timer_->disableTimer(); }
+
+void ScopedRdsConfigSubscription::cleanupDefferedDeletes() {
+  for (auto iter = deffered_to_be_deleted_.begin(); iter != deffered_to_be_deleted_.end();) {
+    if (iter->use_count() == 1) {
+      iter = deffered_to_be_deleted_.erase(iter);
+    } else {
+      iter++;
+    }
+  }
+  if (!deffered_to_be_deleted_.empty()) {
+    // Cleanup as fast as possible.
+    cleanup_timer_->enableTimer(std::chrono::milliseconds(100));
+  }
+}
+
 void ScopedRdsConfigSubscription::onConfigUpdate(
-    const Protobuf::RepeatedPtrField<ProtobufWkt::Any>& resources,
+    const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources,
+    const Protobuf::RepeatedPtrField<std::string>& removed_resources,
     const std::string& version_info) {
-  std::vector<envoy::api::v2::ScopedRouteConfiguration> scoped_routes;
-  for (const auto& resource_any : resources) {
-    scoped_routes.emplace_back(MessageUtil::anyConvert<envoy::api::v2::ScopedRouteConfiguration>(
-        resource_any, validation_visitor_));
+  bool any_applied = false;
+  std::vector<std::string> exception_msgs;
+  absl::flat_hash_set<std::string> unique_resource_names;
+  envoy::config::filter::network::http_connection_manager::v2::Rds rds;
+  rds.mutable_config_source()->MergeFrom(rds_config_source_);
+
+  // If new route config sources come after the factory_context_.initManager()'s initialize() been
+  // called, that initManager can't accept new targets. Instead we use a local override which will
+  // start new subscriptions but not wait on them to be ready.
+  std::unique_ptr<Init::ManagerImpl> overriding_init_manager;
+  if (factory_context_.initManager().state() == Init::Manager::State::Initialized) {
+    // Pause RDS to not send a burst of RDS requests until we start all the new subscriptions.
```

Comment:
Why only when already initialized? Shouldn't we always pause? Even we don't have to from a wire perspective (e.g. the RDS providers aren't sending messages yet or something), I think conceptually i'ts clearer to always pause/resume around this processing.

Reply:
Do you have some code links for the assertion in the first paragraph? This seems quite fragile, relying on this action at a distance. Is there a way to be more robust in a way that is locally easier to reason about?